lua/rollingLimit.lua.json

```diff
@@ -1 +1 @@
 then -- this is a corruption\n      -- we make up a time that would fill this limit via addTokens below\n      lastUpdateMS = nowMS - ((prevTokens / limit) * intervalMS)\n   else\n      lastUpdateMS = initialUpdateMS\n   end\nend\n\n-- tokens that should have been added by now\n-- note math.max in case this ends up negative (clock skew?)\n-- now that we call 'TIME' this is less likely to happen\nlocal addTokens = math.max(((nowMS - lastUpdateMS) / intervalMS) * limit, 0)\n\n-- calculated token balance coming into this transaction\nlocal grossTokens = math.min(prevTokens + addTokens, limit)\n\n-- token balance after trying this transaction\nlocal netTokens = grossTokens - amount\n\n-- time to fill enough to retry this amount\nlocal retryDelta = 0\n\nlocal rejected = false\nlocal forced = false\n\n-- Whether this withdrawal is greater than our limit\nlocal largerWithdrawal = false\n\nif netTokens < 0 then -- we used more than we have\n   if force then\n      forced = true\n      netTokens = 0 -- drain the swamp\n   elseif allowLargerWithdrawal and grossTokens == limit then\n     -- If allowLargerWithdrawal, we do not reject and we allow\n     -- net tokens to remain below zero IFF the bucket is full\n     -- (eg grossTokens == limit)\n     largerWithdrawal = true\n   else\n      rejected = true\n      netTokens = grossTokens -- rejection doesn't eat tokens\n   end\n   -- == percentage of `intervalMS` required before you have `amount` tokens\n   retryDelta = math.ceil(((amount - netTokens) / limit) * intervalMS)\n\n   -- If this is a largerWithdrawal the retry delta is only the amount of time it\n   -- takes to refill the bucket again\n   if largerWithdrawal then\n      retryDelta = math.ceil(((limit - netTokens) / limit) * intervalMS)\n   end\nelse -- polite transaction\n   -- nextNet == pretend we did this again...\n   local nextNet = netTokens - amount\n   if nextNet < 0 then -- ...we would need to wait to repeat\n      -- == percentage of `invervalMS` required before you would have `amount` tokens again\n      retryDelta = math.ceil((math.abs(nextNet) / limit) * intervalMS)\n   end\nend\n\n-- rejected requests don't cost anything, we'll wait for a costly request to update our values\n-- forced requests show up here as !rejected, but with netTokens = 0 (drained)\nif rejected == false then\n\n   local expirationMS = intervalMS\n\n   if allowLargerWithdrawal then\n      expirationMS = math.ceil(((limit - netTokens) / limit) * intervalMS)\n   end\n\n   redis.call('PSETEX',valueKey,expirationMS,netTokens)\n\n   if addTokens > 0 or initialUpdateMS == false then\n      -- we filled some tokens, so update our timestamp\n      redis.call('PSETEX',timestampKey,expirationMS,nowMS)\n   else\n      -- we didn't fill any tokens, so just renew the timestamp so it survives with the value\n      redis.call('PEXPIRE',timestampKey,expirationMS)\n   end\nend\n\nreturn { netTokens, rejected, retryDelta, forced }\n","sha1":"3079e799adf1c854cdbeb0ef6a55953db3fb2b75"}
```

Comment:
Is this a stringified copy of `rollingLimit.lua`, in source control?

Reply:
Yeah I believe they check it in so that they can precompute the sha hash and then they can publish exactly the code that's in the repo (instead of having some npm build step that produces this artifact). The sha has is part of how redis caches scripts that are run. I'm sort of ambivalent about this approach vs others, but I think it basically works fine and I don't want to make significant changes to this repo for now in hopes that we can upstream this and not have to maintain a fork.